Use Attoparsec for parsing git output instead of manually splitting Text

lalaithion opened the pull request:
Also added tests for parsing git output.

Differences in behavior: File paths can now have any character except for the null character in them, where previously they could not contain spaces, tabs or the null character.

Fixes #34

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -30,16 +36,45 @@ catFile gitDir (OID oid) = sh $ do
+  pure $ nullTreeEntry
+
+-- | Attoparsec parser for a single line of git ls-tree -rz output
+entrySuccessParser :: Parser TreeEntry
```
Please use applicative style for this parser. [This parser for Core](https://github.com/github/semantic/blob/master/semantic-core/src/Data/Core/Parser.hs#L69-L74) is a good place to take inspiration from. I also think you can simplify this: I’m not sure you need the `takeWhileToNul` helper, since NULs are guaranteed not to exist except at the end of a record.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -30,16 +36,45 @@ catFile gitDir (OID oid) = sh $ do
+  oid <- takeWhileToNul (/= '\t')
+  _ <- AP.char '\t'
+  path <- takeWhileToNul (const True)
+  pure $ TreeEntry (objectMode mode) (objectType ty) (OID oid) (unpack path)
```
The `objectMode`, `objectType`, etc. functions (in `Semantic.Git`) are the functions I wanted to get rid of with this parser (as they are only used in parsing.) Can you write a parser that expresses them more directly? e.g.

```
parseObjectType :: Parser ObjectType
parseObjectType = choice [ BlobObject <$ “blob”, TreeObject <$ “tree”]
```

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -30,16 +36,45 @@ catFile gitDir (OID oid) = sh $ do
+
+-- | Attoparsec parser for a block af text ending with \NUL
+-- in order to consume invalid input
+entryDefaultParser :: Parser TreeEntry
```
If we hit invalid input while parsing git-ls-tree output, I think we can and should bail, as it means either `git` or our code is very screwed up. I wouldn’t use a default parser; failure should be reified in the act of parser failure itself, if that makes sense.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -30,16 +36,45 @@ catFile gitDir (OID oid) = sh $ do
 lsTree :: FilePath -> OID -> IO [TreeEntry]
 lsTree gitDir (OID sha) = sh $ do
   out <- run "git" [pack ("--git-dir=" <> gitDir), "ls-tree", "-rz", sha]
```
```suggestion
  parseEntries <$> run "git" [pack ("--git-dir=" <> gitDir), "ls-tree", "-rz", sha]
```

patrickt commented on test/Semantic/Spec.hs:
```diff
@@ -24,5 +25,22 @@ spec = parallel $ do
+      let expected = TreeEntry NormalMode TreeObject (OID "ThisIsTheOid") "/this/is/the/path"
+      parseEntry input `shouldBe` expected
+
+    it "parses nonsense into a default value" $ do
```
I don’t think we need this one, though: if we get nonsense I want the parser to complain loudly, and we can handle failure at a higher-level.

patrickt commented on test/Semantic/Spec.hs:
```diff
@@ -24,5 +25,22 @@ spec = parallel $ do
+      let expected = TreeEntry OtherMode OtherObjectType (OID mempty) mempty
+      parseEntry input `shouldBe` expected
+
+    it "parses many outputs separated by \\NUL" $ do
```
I’d also like to see an integration test that used `git` invocations in a temporary directory, but I think I’ll take care of that in a separate PR.

patrickt replied:
Turns out I already did this and completely forgot about it.

patrickt commented on test/Semantic/Spec.hs:
```diff
@@ -24,5 +25,22 @@ spec = parallel $ do
       output `shouldBe` "(Statements\n  (Method\n    (Empty)\n    (Identifier)\n    (Statements)))\n"
+
+  describe "gitParsing" $ do
+    it "parses a git output string" $ do
```
Great job on the tests!

patrickt commented on test/Semantic/Spec.hs:
```diff
@@ -24,5 +25,22 @@ spec = parallel $ do
       output <- fmap runBuilder . runTaskOrDie $ parseTermBuilder TermSExpression [methodsBlob]
       output `shouldBe` "(Statements\n  (Method\n    (Empty)\n    (Identifier)\n    (Statements)))\n"
+
+  describe "gitParsing" $ do
```
```suggestion
  describe “git ls-tree parsing" $ do
```

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -30,16 +36,45 @@ catFile gitDir (OID oid) = sh $ do
 
+-- | Parses an list of entries separated by \NUL, and on failure return []
+parseEntries :: Text -> [TreeEntry]
+parseEntries = either (const []) id . AP.parseOnly (AP.sepBy entryParser (AP.char '\NUL'))
```
We want to parse the entire input, so you should probably do
`AP.parseOnly (entries <* eof) where entries = entryParser `sepBy` AP.char '\NUL'`

lalaithion commented:
I've tried to follow most of the comments above in the most recent commit. In particular:

1) I've rewritten all of the functions to use Attoparsec
2) I've used applicative style
3) I've made the parser more likely to fail instead of falling back on defaults. In particular, OtherObjectType and OtherMode are no longer created by the parser if the Mode or Type is unknown, and the OID is only allowed to be lowercase hexadecimal.
4) I've added a test to ensure whitespace can be in filenames
5) I've manually tested `cabal new-run semantic -- parse --gitDir=<...> --sha=HEAD --json` on many local git repositories. In order for this to work, I had to make a few other changes to the code, as the current git invocation ran into the issue described here: https://stackoverflow.com/questions/1386291/git-git-dir-not-working-as-expected

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
 catFile :: FilePath -> OID -> IO Text
 catFile gitDir (OID oid) = sh $ do
-  run "git" [pack ("--git-dir=" <> gitDir), "cat-file", "-p", oid]
+  run "git" ["-C", pack gitDir, "cat-file", "-p", oid]
```
Didn't know about this! Looks like it landed a few dot-releases ago, so this is a 👍 for me.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
 
+-- | Parses an list of entries separated by \NUL, and on failure return []
+parseEntries :: Text -> [TreeEntry]
+parseEntries text = case parseOnly everything text of
```
I think this can be a good deal simpler. Pull in the `parsers` library and use this.

```
parseOnly ((entryParser `sepEndBy` char '\NUL') <* endOfInput)
```

You shouldn't need to call `feed` or override `parseOnly`.

lalaithion replied:
`parseOnly` from Attoparsec? Or from another library? I was using this approach because `Data.Attoparsec.Text.parseOnly` returns minimal error information, which I ran into while debugging my parser. I'm happy to do this, but it will further degrade errors.

patrickt replied:
Remove it for now, and if it becomes a problem we can switch to Megaparsec or Trifecta (which are slower but have way better messaging).

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+parseEntries :: Text -> [TreeEntry]
+parseEntries text = case parseOnly everything text of
+  Done "" ls -> ls
+  other      -> error ("There was an error parsing the Git output: " <> show other)
```
Doing the above should obviate this `error` call (which should just return `[]`).

patrickt replied:
I agree, however, that we need a better error handling story.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+-- | Parse the entire input with entryParser, and on failure return a default
+-- For testing purposes only
+parseEntry :: Text -> TreeEntry
+parseEntry = either (const nullTreeEntry) id . AP.parseOnly (entryParser <* AP.endOfInput)
```
I want to kill `nullTreeEntry`, so can you make this `Text -> Either String TreeEntry`?

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+-- | Parses a TreeEntry
+entryParser :: Parser TreeEntry
+entryParser = TreeEntry
+  <$> modeParser <* (" " <?> "First Space")
```
```suggestion
  <$> modeParser <* char ' '
```

Error messages with attoparsec are mostly a lost cause, but there's no reason to annotate a space with an error message, given that the error message will say it expected a space anyway. I also think that using the `char` parser here is faster and a little more explicit than relying on `OverloadedStrings` here.

lalaithion replied:
When testing this code, I ended up debugging a parse output that failed on the parser "string", with no further context. Adding these annotations made me able to figure out which parser was failing. I'm still happy to remove them, but it will further degrade error messages.

patrickt replied:
Makes sense. I don’t think we really need to worry about error messages for now, given the relatively normal shape of this data, now that you’ve debugged it, so I’d remove the annotations for the sake of readability. (`char ‘ ‘` will give a better error message than `string “ “`, by the way).

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+entryParser :: Parser TreeEntry
+entryParser = TreeEntry
+  <$> modeParser <* (" " <?> "First Space")
+  <*> typeParser <* (" " <?> "Second Space")
```
Ditto.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+entryParser = TreeEntry
+  <$> modeParser <* (" " <?> "First Space")
+  <*> typeParser <* (" " <?> "Second Space")
+  <*> (OID <$> AP.takeWhile (AP.inClass "0123456789abcdef") <?> "OID Parser") <* ("\t" <?> "Tab")
```
Two things: pull this out into a top-level function (or down to the `where` clause), and you can also use `satisfy Data.Char.isHexDigit`. No need for `<?>` annotations here either.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+  <$> modeParser <* (" " <?> "First Space")
+  <*> typeParser <* (" " <?> "Second Space")
+  <*> (OID <$> AP.takeWhile (AP.inClass "0123456789abcdef") <?> "OID Parser") <* ("\t" <?> "Tab")
+  <*> (unpack <$> AP.takeWhile (/= '\NUL') <?> "Filepath") <?> "Entry Parser"
```
`manyTill (== '\NUL')` might avoid the overhead of an `unpack` here.

lalaithion replied:
The documentation claims that "takeWhile" is a more efficient parser than "manyTill", so I'm not sure about this one.

patrickt commented on src/Semantic/Git.hs:
```diff
@@ -24,22 +30,40 @@ clone url path = sh $ do
+  <*> (OID <$> AP.takeWhile (AP.inClass "0123456789abcdef") <?> "OID Parser") <* ("\t" <?> "Tab")
+  <*> (unpack <$> AP.takeWhile (/= '\NUL') <?> "Filepath") <?> "Entry Parser"
+    where
+      typeParser = AP.choice [BlobObject <$ "blob", TreeObject <$ "tree"] <?> "Type Parser"
```
No need for annotations here either.

patrickt commented on test/Semantic/Spec.hs:
```diff
@@ -24,5 +25,22 @@ spec = parallel $ do
+
+    it "parses many outputs separated by \\NUL" $ do
+      let input = "100644 tree abcdef\t/this/is/the/path\NUL120000 blob 17776\t/dev/urandom\NUL\n"
+      let expected = [ TreeEntry NormalMode TreeObject (OID "abcdef") "/this/is/the/path", TreeEntry SymlinkMode BlobObject (OID "17776") "/dev/urandom"]
```
Is this a 17776 reference (by Jon Bois)? Love it.

lalaithion replied:
Yes, it is.
